### DISCUSSION TITLE
`jsconfig.json` contrary to documentation implies `maxNodeModuleJsDepth: 2` etc

### DISCUSSION THREAD

**Discussion by voxpelli** [voxpelli/types-in-js#25] [General]
The only documentation of what additional rules `jsconfig.json` sets that I can find are the one in [the VSCode docs](https://code.visualstudio.com/docs/languages/jsconfig), and it says (emphasis mine):

> **Tip:** `jsconfig.json` is a descendant of [tsconfig.json](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html), which is a configuration file for TypeScript. **`jsconfig.json` is `tsconfig.json` with `"allowJs"` attribute set to `true`.**

But, turns out that it hasn't been true for 7.5 years. Here's a list of all additions on top of `allowJs: true` that has happened:

* **Aug 25, 2016:** When `maxNodeModuleJsDepth` was changed to `0` as default for TS as a whole, it was kept as `2` for `jsconfig.json` (without updating the docs it seems) https://github.com/microsoft/TypeScript/commit/c0309fa78e5dbf7dd699cf21ea59ed9cb24268ad / https://github.com/microsoft/TypeScript/pull/10538
* **Sep 23, 2016:** `allowSyntheticDefaultImports: true` was added in https://github.com/microsoft/TypeScript/commit/12578a5cc6f2d9542bea1f3f47c8475d8c1d129a / https://github.com/microsoft/TypeScript/pull/11070
* **Oct 12, 2016:** `skipLibCheck: true` was added in https://github.com/microsoft/TypeScript/commit/460de66311b26457481a6d2055231e00fa24e705 / https://github.com/microsoft/TypeScript/pull/11399
* **Jan 25, 2018:** ` noEmit: true` was added in https://github.com/microsoft/TypeScript/commit/4af47f188772591ecd996eb18a853c2cf1cd83f2 / https://github.com/microsoft/TypeScript/pull/21505

And that's where we [remain today](https://github.com/microsoft/TypeScript/blob/2d70b57df4b64a3daef252abb014562e6ccc8f3c/src/compiler/commandLineParser.ts#L3433) with `jsconfig.json` implying:

```json
{
  "allowJs": true,
  "maxNodeModuleJsDepth": 2,
  "allowSyntheticDefaultImports": true,
  "skipLibCheck": true,
  "noEmit": true
}
```

It seems that the VSCode docs are long overdue for an update: https://github.com/microsoft/vscode-docs/blame/91835d57149c4352073c22300487a18baeefe13c/docs/languages/jsconfig.md#L18 (Which I will do following this post)

And likewise the documentation of [`maxNodeModuleJsDepth`](https://www.typescriptlang.org/tsconfig/#maxNodeModuleJsDepth) also needs to be updated:

> Ideally this should stay at 0 (the default), and `d.ts` files should be used to explicitly define the shape of modules. However, there are cases where you may want to turn this on at the expense of speed and potential accuracy.

0 is not the default for `jsconfig.json` files, it defaults to 2.

And since `maxNodeModuleJsDepth` is only usable when `allowJs` is `true` it is quite likely that someone looking at that rule is also thinking about `jsconfig.json`.

---

Based on a realization in https://github.com/eslint/eslint/issues/18100#issuecomment-1971500684, initially posted as a comment in https://github.com/voxpelli/types-in-js/discussions/2#discussioncomment-8633488

**Comment by voxpelli** [comment]
TLDR is essentially: Don't use `jsconfig.json`, its an oddity that sets discouraged options like [`maxNodeModuleJsDepth`](https://www.typescriptlang.org/tsconfig/#maxNodeModuleJsDepth) without even documenting it.

That in turn can cause:

- False errors – as shown by eg https://github.com/eslint/eslint/issues/18100
- Slow performance – as all the JS two levels down in `maxNodeModuleJsDepth` will get parsed if `d.ts` files are missing for them
- Inconsistent type checking – if a module appears deeper down in `node_modules` than 2 steps, then it won't be picked up as `maxNodeModuleJsDepth` is only set to 2 (a sensible value for those cases where one need to use it, a case which existed before one could generate `d.ts` files from ones JSDoc comments, now that's the preferable solution)

Other defaults that can cause issues:

- `skipLibCheck` – disables all type checking of `d.ts` files, not only the type checking of `d.ts` files in `node_modules`. That means that if you do write any `d.ts` files of your own in your project, the `jsconfig.json` defaults to disabling them. See issues it can cause here: https://github.com/sindresorhus/type-fest/pull/785 And me choosing to have it disabled here: https://github.com/voxpelli/tsconfig/issues/1

Worth noting:

- `allowSyntheticDefaultImports` – will already be default when `"module": "node16"`  or `"module": "nodenext"`, something most JS-projects should have set anyways (and something I have in [my base config](https://github.com/voxpelli/tsconfig/blob/44dac2276400d1be2840fee31c94df44e7bf2749/base.json#L6-L7)). It's also active in eg [`@tsconfig/recommended`](https://www.npmjs.com/package/@tsconfig/recommended) as it sets `esModuleInterop` which in turns sets `allowSyntheticDefaultImports` and  `"module": "node16"`  or `"module": "nodenext"` sets `esModuleInterop`

Options that the default doesn't set and which you should set:

- `strict` – always always always set `strict: true`

For more suggestions, see my base config, which also links to other base configs: https://github.com/voxpelli/tsconfig My base config is set up purely for a types in JS workflow (backed up by occasional manual `d.ts`)
